backend/src/SerialHandler.py

```diff
@@ -0,0 +1,261 @@
+
+        encBuf = Codec.Encode(buf, len(buf), ProtoCore.MessageID.MSG_COMMAND)
+        target_enum = utl.get_node_from_str(target)
+        if (target_enum == ProtoCore.NODE_DMB or target_enum == ProtoCore.Node.NODE_PBB) and self.port == RADIO_SERIAL_PORT:
+            self.serial_port.write(encBuf)
+        if (target_enum ==  ProtoCore.NODE_RCU or target_enum ==  ProtoCore.Node.NODE_SOB) and self.port == UART_SERIAL_PORT:
+            self.serial_port.write(encBuf)
+
+        return True
+
+# Procedures =======================================================================================
+
+def process_serial_workq_message(message: WorkQ_Message, ser_han: SerialHandler) -> bool:
+        """
+        Process the message from the workq.
+
+        Args:
+            message (str):
+                The message from the workq.
+        """
+        logger.debug(f"Processing serial workq message: {message.message_type}")
+        messageID = message.message_type
+
+        if messageID == THREAD_MESSAGE_KILL:
+            logger.debug(f"Killing {ser_han.thread_name} thread")
+            return False
+        elif messageID == THREAD_MESSAGE_SERIAL_WRITE:
+            command = message.message[0]
+            target = message.message[1]
+            command_param = message.message[2]
+            source_sequence_number = message.message[3]
+            ser_han.send_serial_command_message(command, target, command_param, source_sequence_number)
+        elif messageID == TEST_RECEIVE_SERIAL_MESSAGE:
+            ser_han.handle_serial_message(message.message[0])
+        return True
+
+def serial_thread(thread_name: str, device: SerialDevices, baudrate: int, thread_workq: mp.Queue, message_handler_workq: mp.Queue):
+    """
+    Thread function for the incoming serial data listening.
+
+    Args:
+        thread_name (str):
+            The name of the thread.
+        device (SerialDevices):
+            The device to listen to.
+        baudrate (int):
+            The baudrate of the serial connection.
+        thread_workq (mp.Queue):
+            The workq to send messages to the thread.
+        message_handler_workq (mp.Queue):
+            The workq to send messages to the database handler. 
+    """
+    if device == SerialDevices.UART:
+        port = UART_SERIAL_PORT
+    elif device == SerialDevices.RADIO:
+        port = RADIO_SERIAL_PORT
+    
+    serial_workq = thread_workq
+    ser_han = SerialHandler(thread_name, port, baudrate, message_handler_workq)
+    if ser_han.serial_port == None:
+        return
+
+    while (1):
+        # If there is any workq messages, process them first
+        # then once the queue is empty read the serial port
+        if not serial_workq.empty():
+            if not process_serial_workq_message(serial_workq.get(), ser_han):
+                return
+        else:
+            ser_han.handle_serial_message()
+
+        time.sleep(0.1)
```

Comment:
This sleep 0.1 might be too much of a delay (100ms between each serial message read, correct me if wrong) -- while PySerial is buffered (which is probably why there aren't too many lost messages with just the RCU) it technically wouldn't keep up with a rate of > 10 messages per second right now.

There's also a weird dynamic here where read_until blocks until an incoming message is processed so our commands to write serial would probably stop working if we stopped receiving data for some reason.

One way to deal with this might be to use threading inside this process to make one separate WorkQ thread that just blocks forever for a workQ message, and the serial Rx thread which will block in read_until for a serial message, so both Tx/Rx will work at the rate that messages come in, and we won't have to poll at a fixed rate.

Reply:
I added this, needed to be veerified on hardware but software wise seems fine